File: rust/candid_parser/tests/assets/ok/service.js
import { IDL } from '@dfinity/candid';

export const Service = IDL.Rec();
export const Func = IDL.Func([], [Service], []);
Service.fill(IDL.Service({ 'f' : Func }));
export const Service2 = Service;

export const idlService = IDL.Service({
  'asArray' : IDL.Func([], [IDL.Vec(Service2), IDL.Vec(Func)], ['query']),
  'asPrincipal' : IDL.Func([], [Service2, Func], []),
  'asRecord' : IDL.Func([], [IDL.Tuple(Service2, IDL.Opt(Service), Func)], []),
  'asVariant' : IDL.Func(
      [],
      [
        IDL.Variant({
          'a' : Service2,
          'b' : IDL.Record({ 'f' : IDL.Opt(Func) }),
        }),
      ],
      [],
    ),
});

export const idlInitArgs = [];

/**
 * @deprecated Import IDL types directly from this module instead of using this factory function.
 */
export const idlFactory = ({ IDL }) => {
  const Service = IDL.Rec();
  const Func = IDL.Func([], [Service], []);
  Service.fill(IDL.Service({ 'f' : Func }));
  const Service2 = Service;
  return IDL.Service({
    'asArray' : IDL.Func([], [IDL.Vec(Service2), IDL.Vec(Func)], ['query']),
    'asPrincipal' : IDL.Func([], [Service2, Func], []),
    'asRecord' : IDL.Func(
        [],
        [IDL.Tuple(Service2, IDL.Opt(Service), Func)],
        [],
      ),
    'asVariant' : IDL.Func(
        [],
        [
          IDL.Variant({
            'a' : Service2,
            'b' : IDL.Record({ 'f' : IDL.Opt(Func) }),
          }),
        ],
        [],
      ),
  });
};

/**
 * @deprecated Import IDL types directly from this module instead of using this factory function.
 */
export const init = ({ IDL }) => { return []; };

File: rust/candid_parser/tests/assets/ok/service.rs
// This is an experimental feature to generate Rust binding from Candid.
// You may want to manually adjust some of the types.
#![allow(dead_code, unused_imports)]
use candid::{self, CandidType, Deserialize, Principal};
use ic_cdk::call::{Call, CallResult as Result};

candid::define_function!(pub Func : () -> (Service));
candid::define_service!(pub Service : { "f" : Func::ty() });
pub type Service2 = Box<Service>;
#[derive(CandidType, Deserialize)]
pub enum AsVariantRet {
  #[serde(rename="a")]
  A(Service2),
  #[serde(rename="b")]
  B{ f: Option<Func> },
}

pub struct Service(pub Principal);
impl Service {
  pub async fn as_array(&self) -> Result<(Vec<Service2>,Vec<Func>,)> {
    Ok(Call::bounded_wait(self.0, "asArray").await?.candid()?)
  }
  pub async fn as_principal(&self) -> Result<(Service2,Func,)> {
    Ok(Call::bounded_wait(self.0, "asPrincipal").await?.candid()?)
  }
  pub async fn as_record(&self) -> Result<((Service2, Option<Service>, Func),)> {
    Ok(Call::bounded_wait(self.0, "asRecord").await?.candid()?)
  }
  pub async fn as_variant(&self) -> Result<(AsVariantRet,)> {
    Ok(Call::bounded_wait(self.0, "asVariant").await?.candid()?)
  }
}
/// Canister ID: `aaaaa-aa`
pub const CANISTER_ID : Principal = Principal::from_slice(&[]);
pub const service : Service = Service(CANISTER_ID);

